openvdb/points/PointScatter.h

```diff
@@ -0,0 +1,676 @@
+        const Index64 voxelCount = leafManager.activeLeafVoxelCount();
+        assert(voxelCount != 0);
+
+        const double pointsPerVolume = double(mPointCount) / double(voxelCount);
+        mPointsPerVoxel = static_cast<Index32>(math::RoundDown(pointsPerVolume));
+        Index64 remainder = mPointCount - (mPointsPerVoxel * voxelCount);
+
+        if (remainder == 0) {
+            UniformVoxelPointScatter<RandGenT, ValueType, CodecT, PointDataGridT, InterruptT>
+                op(float(mPointsPerVoxel), BaseT::mSeed, BaseT::mSpread, BaseT::mInterrupter);
+            op(grid);
+            BaseT::mPoints = op.points();
+            return;
+        }
+
+        std::vector<Index64> voxelOffsets;
+        std::thread worker(&getPrefixSum, std::ref(leafManager), std::ref(voxelOffsets));
+
+        math::RandInt<Index64, RandGenT> gen(BaseT::mSeed, 0, voxelCount-1);
+        std::vector<Index64> values;
+        values.reserve(remainder);
+        for (Index64 i = 0; i < remainder; ++i) values.emplace_back(gen());
+        tbb::parallel_sort(values.begin(), values.end());
+
+        worker.join();
+
+        leafManager.foreach([&](typename BaseT::LeafNodeT& leaf, const size_t idx) {
+            const Index64 lowerOffset = voxelOffsets[idx]; // inclusive
+            const Index64 upperOffset = voxelOffsets[idx + 1]; // exclusive
+            assert(upperOffset > lowerOffset);
+
+            const auto valuesEnd = values.end();
+            auto lower = std::lower_bound(values.begin(), valuesEnd, lowerOffset);
+
+            auto* const data = leaf.buffer().data();
+            auto iter = leaf.beginValueOn();
+
+            Index32 currentOffset(0);
+            while (lower != valuesEnd) {
+                const Index64 vId = *lower;
+                if (vId >= upperOffset) break;
+
+                const Index32 nextOffset = Index32(vId - lowerOffset);
+                iter.increment(nextOffset - currentOffset);
+                currentOffset = nextOffset;
+                assert(iter);
+
+                auto& value = data[iter.pos()];
+                value = value + 1; // no += operator support
+                ++lower;
+            }
+        });
+
+        leafManager.foreach([&](typename BaseT::LeafNodeT& leaf, const size_t) {
+            if (BaseT::wasInterrupted()) return;
+            Index32 offset(0);
+            for (auto iter = leaf.beginValueAll(); iter; ++iter) {
+                if (iter.isValueOn()) {
+                    const Index32 value = mPointsPerVoxel + Index32(*iter);
+                    if (value == 0) leaf.setValueOff(iter.pos());
+                    else            offset += value;
+                }
+                // @note can't use iter.setValue(offset) on point grids
+                leaf.setOffsetOnly(iter.pos(), offset);
+            }
+
+            if (offset != 0) BaseT::generatePositions(leaf, offset);
+        }, /*threaded*/false);
```

Comment:
Is there any advantage in putting this in a foreach lambda, if you're just going to disable the threading?

Reply:
No and actually I don't need to construct a leaf manager anymore in a few cases - I'll rework this.